velox/exec/Spiller.cpp

```diff
@@ -0,0 +1,402 @@
+// sorted spills since for hash join spilling we just use the data in
+// the RowContainer as is.
+class RowContainerSpillStream : public SpillStream {
+ public:
+  RowContainerSpillStream(
+      RowTypePtr type,
+      int32_t numSortingKeys,
+      memory::MemoryPool& pool,
+      Spiller::SpillRows&& rows,
+      Spiller& spiller)
+      : SpillStream(std::move(type), numSortingKeys, pool),
+        rows_(std::move(rows)),
+        spiller_(spiller) {
+    if (!rows_.empty()) {
+      nextBatch();
+    }
+  }
+
+  uint64_t size() const override {
+    return 0;
+  }
+
+ private:
+  void nextBatch() override {
+    // Extracts up to 64 rows at a time. Small batch size because may
+    // have wide data and no gain in larger.when the caller will go
+    // over aggregations row by row.
+    static constexpr vector_size_t kMaxRows = 64;
+    constexpr uint64_t kMaxBytes = 4 << 20;
+    size_t bytes = 0;
+    vector_size_t numRows = 0;
+    auto limit = std::min<size_t>(rows_.size() - nextBatchIndex_, kMaxRows);
+    assert(!rows_.empty());
+    for (; numRows < limit; ++numRows) {
+      bytes += spiller_.container().rowSize(rows_[nextBatchIndex_ + numRows]);
+      if (bytes > kMaxBytes) {
+        ++numRows;
+        break;
+      }
+    }
+    spiller_.extractSpill(
+        folly::Range(&rows_[nextBatchIndex_], numRows), &rowVector_);
+    nextBatchIndex_ += numRows;
+    size_ = rowVector_->size();
+    index_ = 0;
+  }
+
+  Spiller::SpillRows rows_;
+  Spiller& spiller_;
+  size_t nextBatchIndex_ = 0;
+};
+} // namespace
+
+std::unique_ptr<SpillStream> Spiller::spillStreamOverRows(int32_t partition) {
+  VELOX_CHECK(spillFinalized_);
+  VELOX_CHECK_LT(partition, spillRuns_.size());
+  ensureSorted(spillRuns_[partition]);
+  return std::make_unique<RowContainerSpillStream>(
+      rowType_,
+      container_.keyTypes().size(),
+      pool_,
+      std::move(spillRuns_[partition].rows),
+      *this);
+}
+
+void Spiller::ensureSorted(SpillRun& run) {
+  if (!run.sorted) {
+    std::sort(
+        run.rows.begin(),
+        run.rows.end(),
+        [&](const char* left, const char* right) {
+          return container_.compareRows(left, right) < 0;
+        });
+    run.sorted = true;
+  }
+}
+
+std::unique_ptr<Spiller::SpillStatus> Spiller::writeSpill(
+    int32_t partition,
+    uint64_t maxBytes) {
+  // Size Target size of a single vector of spilled content. One of
+  // these will be materialized at a time for each stream of the
+  // merge.
+  constexpr int32_t kTargetBatchBytes = 1 << 20; // 1MB
+
+  RowVectorPtr spillVector;
+  auto& run = spillRuns_[partition];
+  try {
+    ensureSorted(run);
+    int64_t totalBytes = 0;
+    int32_t written = 0;
+    while (written < run.rows.size()) {
+      int32_t i = 0;
+      int32_t limit = std::min<uint64_t>(128, run.rows.size() - written);
+      int32_t bytes = 0;
+      for (; i < limit; ++i) {
+        bytes += container_.rowSize(run.rows[written + i]);
```

Comment:
this logic repeats in nextBatch; would it be possible to extract a helper method to avoid copy-paste?